### PR TITLE
FE: Mark Fields as hidden in parameter mapping

### PR DESCRIPTION
**Pull request by uladzimirdev**
Epic: https://github.com/metabase/metabase/issues/30016

Show a badge and handle errors in mapping form

Demo https://www.loom.com/share/3cb103a71d0c44c382a370b47ddeaa8d

### Additional work
- [x] add unit tests
- [x] replace connect with useDispatch
- [x] Simplify ActionFormMapper component and move state up
- [x] e2e test which covers showing hidden badge, error and state of "done" button

### CODE REVIEW

File: frontend/src/metabase/actions/components/ActionViz/ActionParameterMapper.tsx
Comment (kamilmielnik) [inline frontend/src/metabase/actions/components/ActionViz/ActionParameterMapper.tsx:97]: The first thing I noticed is that the badge is unusually close to the text.
I don't think that `{" "}` space is enough of a visual separator and I think we're going to need CSS here.

In Figma there is 8px of space between the badge and text and it looks much more natural:

image

vs here:

image

File: frontend/src/metabase/actions/components/ActionViz/ActionParameterMapper.tsx
Comment (kamilmielnik) [inline frontend/src/metabase/actions/components/ActionViz/ActionParameterMapper.tsx:85]: 4 optional chains is an obvious code smell and generates questions:
1. Why can an `action` be `undefined` in a `ActionParameterMapper`? If it can't, we should make code communicate this. For example we could do `if (typeof action === "undefined") { throw new Error("..."); }` in this component.
2. Are there any other places where we need to get so deep? Perhaps there are some utilities that could be reused and/or refactored to reduce the clutter. How about [`useActionForm`](https://github.com/metabase/metabase/blob/master/frontend/src/metabase/actions/hooks/use-action-form/use-action-form.ts#L25)?

File: frontend/src/metabase/actions/components/ActionViz/ActionParameterMapper.tsx
Comment (uladzimirdev) [reply frontend/src/metabase/actions/components/ActionViz/ActionParameterMapper.tsx:97]: nice catch, I'll fix it!

File: frontend/src/metabase/actions/components/ActionViz/ActionParameterMapper.tsx
Comment (uladzimirdev) [reply frontend/src/metabase/actions/components/ActionViz/ActionParameterMapper.tsx]: I simplified expressions with introducing two separate helpers to get required and hidden states of the parameter.

As for general improvement of action existence, I'm not confident enough to make such refactoring right now

File: frontend/src/metabase/actions/components/ActionViz/ActionParameterMapper.tsx
Comment (uladzimirdev) [reply frontend/src/metabase/actions/components/ActionViz/ActionParameterMapper.tsx]: should be fixed now. I think the space is too large btw

File: frontend/src/metabase/actions/components/ActionViz/ActionDashcardSettings.unit.spec.tsx
Comment (ranquild) [reply frontend/src/metabase/actions/components/ActionViz/ActionDashcardSettings.unit.spec.tsx:210]: Yeah I'd use plain strings, catching string changes can be a good thing sometimes

File: frontend/src/metabase/actions/components/ActionViz/ActionParameterMapper.tsx
Comment (ranquild) [inline frontend/src/metabase/actions/components/ActionViz/ActionParameterMapper.tsx:97]: could we use the same text we use for regular invalid form fields? like this 
Screenshot 2023-06-07 at 19 26 04

File: frontend/src/metabase/actions/components/ActionViz/ActionDashcardSettings.unit.spec.tsx
Comment (ranquild) [inline frontend/src/metabase/actions/components/ActionViz/ActionDashcardSettings.unit.spec.tsx:212]: I don't see a test that makes sure you cannot submit the mapping form when one of the fields is required

**Comment by uladzimirdev** [comment]
@ranquild I missed the case we need to prevent form submitting, I'll need to refactor form in this case. Why I missed that - there is no "Cancel" button on the modal and click outside doesn't close the form as well. So the only way to close the modal is to click "Done" /cc @kdoh Does it make sense to add a "Cancel" button maybe?

File: frontend/src/metabase/actions/components/ActionViz/ActionDashcardSettings.unit.spec.tsx
Comment (WiNloSt) [inline frontend/src/metabase/actions/components/ActionViz/ActionDashcardSettings.unit.spec.tsx:108]: nit: I think this is hard to follow what action it's trying to create. I mean the name `actionDashcardWithActionWithHiddenFields` is pretty self-explanatory, but `actions2` is not clear also the magic index number `3`.


do you think it'd be better if you structure variables a bit differently e.g.
#### This is a pseudo code but to demonstrate an idea
```
const actionWithHiddenFields = ... // whatever you put in `actions2[3]`
```
then we could do

```suggestion
  action: actionWithHiddenFields,
```

As a bonus, if `actions2` should have a meaning on its own maybe we could also name it.
```
const actionsForSomething = [actionWithFoo, actionWithBar, actionWithBaz, actionWithHiddenFields]
```

File: frontend/src/metabase/actions/components/ActionViz/ActionDashcardSettings.unit.spec.tsx
Comment (uladzimirdev) [reply frontend/src/metabase/actions/components/ActionViz/ActionDashcardSettings.unit.spec.tsx:224]: added a test

File: frontend/src/metabase/actions/components/ActionViz/ActionDashcardSettings.unit.spec.tsx
Comment (uladzimirdev) [reply frontend/src/metabase/actions/components/ActionViz/ActionDashcardSettings.unit.spec.tsx]: fixed

File: frontend/src/metabase/actions/components/ActionViz/ActionDashcardSettings.unit.spec.tsx
Comment (uladzimirdev) [reply frontend/src/metabase/actions/components/ActionViz/ActionDashcardSettings.unit.spec.tsx:108]: I agree, let me try to come up with some better names

Comment (uladzimirdev) [reply frontend/src/metabase/actions/components/ActionViz/ActionDashcardSettings.unit.spec.tsx]: Fixed only the main place (for `actionDashcardWithActionWithHiddenFields`), as for other places, actions are similar and they are basic, so I can't come up with some good name, probably array structure works the best in this case

File: e2e/test/scenarios/dashboard/actions-on-dashboards.cy.spec.js
Comment (uladzimirdev) [inline e2e/test/scenarios/dashboard/actions-on-dashboards.cy.spec.js:874]: moved `save` to another helper

**Comment by kamilmielnik** [comment]
LGTM but I'll let others approve.

File: frontend/src/metabase/actions/components/ActionViz/ActionParameterMapper.tsx
Comment (uladzimirdev) [reply frontend/src/metabase/actions/components/ActionViz/ActionParameterMapper.tsx]: actually you were right and I improved experience by reworking the component and props, thanks!

File: frontend/src/metabase/actions/components/ActionViz/ActionDashcardSettings.tsx
Comment (ranquild) [inline frontend/src/metabase/actions/components/ActionViz/ActionDashcardSettings.tsx:78]: `!= null` or `0` won't work

Comment (uladzimirdev) [reply frontend/src/metabase/actions/components/ActionViz/ActionDashcardSettings.tsx]: fixed

**Review by ranquild — CHANGES_REQUESTED** [review]
I've investigated this change and it seems that we should test it differently. 

In the form settings modal there is no backend interaction involved, that's why we should test the change using `Jest` tests only. We should be able to add tests for the different cases there - required/non-required params, hidden or not, mapped or not, try different values, including `0`.

For the whole epic given I believe we should have only 2 tests for different action types (basic or query) X places where it can be executed (model details, dashboards, public actions) = 6 tests. We shouldn't be testing validation edge cases in e2e tests here.

Furthermore, the current setup logic in the existing e2e tests is very hard to follow, and I my read is that it's a sign that we use API requests to simulate the app behavior too much. Given that we don't need much e2e tests here, I'd create actions for these e2e tests using the app UI and not API directly. 

Please let me know if you have any questions.

**Comment by uladzimirdev** [comment]
@ranquild 

> Furthermore, the current setup logic in the existing e2e tests is very hard to follow, and I my read is that it's a sign that we use API requests to simulate the app behavior too much. Given that we don't need much e2e tests here, I'd create actions for these e2e tests using the app UI and not API directly.

There is a problem with the logic you provide here. 
1. we have basic e2e tests which cover creating actions through UI, so if some code change broke it, we'll already know it. In my cases I didn't need to verify that action can be created through UI, it means creating them through UI is a waste of time and double testing the same things
2. Setup can be simplified for sure. We can create fixtures which can be reused in many places. E.g. Action with hidden params, regular Action x (query, implicit). Please let me know if this is something you meant with "complicated setup".

> For the whole epic given I believe we should have only 2 tests for different action types (basic or query) X places where it can be executed (model details, dashboards, public actions) = 6 tests. We shouldn't be testing validation edge cases in e2e tests here.

Well, with unit tests I'm not 100% sure I provide all the necessary data and simulate behaviour exactly in the same way as in browser. Nevertheless I appreciate your advice here and I'm going to drop all e2e tests for validation, replace them with integration jest tests where it's possible.

> In the form settings modal there is no backend interaction involved, that's why we should test the change using Jest tests only. We should be able to add tests for the different cases there - required/non-required params, hidden or not, mapped or not, try different values, including 0.

Noted, no interaction with BE - no e2e test 👍

**Comment by uladzimirdev** [comment]
@ranquild added unit tests instead of e2e, please take a look

**Review by ranquild — APPROVED** [review]

**Pull request merged**
